pkg/controller/deployment/deployment_controller.go

```diff
@@ -450,6 +451,11 @@ func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extension
 		return dc.updateDeploymentStatus(allRCs, newRC, deployment)
 	}
 
+	// Cleanup old rcs
+	if deployment.Spec.RevisionHistoryLimit != nil {
```

Comment:
Well, you could check for the replicas of the rcs before trying to delete but ultimately I think I prefer what you have now as it is cheaper than what I am proposing.